OpenRA.Game/Support/Log.cs

```diff
@@ -22,90 +26,147 @@ public struct ChannelInfo
+		static void FlushToDisk()
 		{
-			lock (Channels)
-			{
-				if (Channels.ContainsKey(channelName)) return;
+			foreach (var (_, channel) in Channels)
+				channel.Writer?.Flush();
+		}
 
-				if (string.IsNullOrEmpty(baseFilename))
-				{
-					Channels.Add(channelName, default(ChannelInfo));
-					return;
-				}
+		static void DoWork(object obj)
+		{
+			var token = (CancellationToken)obj;
+			var reader = Channel.Reader;
 
-				foreach (var filename in FilenamesForChannel(channelName, baseFilename))
-					try
-					{
-						var writer = File.CreateText(filename);
-						writer.AutoFlush = true;
-
-						Channels.Add(channelName,
-							new ChannelInfo
-							{
-								Filename = filename,
-								IsTimestamped = isTimestamped,
-								Writer = TextWriter.Synchronized(writer)
-							});
-
-						return;
-					}
-					catch (IOException) { }
-			}
+			while (!token.IsCancellationRequested)
+				if (reader.TryRead(out var item))
+					WriteValue(item);
+
+			while (reader.TryRead(out var item))
+				WriteValue(item);
+
+			FlushToDisk();
 		}
 
-		public static void Write(string channelName, string value)
+		static void WriteValue(ChannelData item)
 		{
-			var channel = Channel(channelName);
+			var channel = GetChannel(item.Channel);
 			var writer = channel.Writer;
 			if (writer == null)
 				return;
 
 			if (!channel.IsTimestamped)
-				writer.WriteLine(value);
+				writer.WriteLine(item.Text);
 			else
 			{
 				var timestamp = DateTime.Now.ToString(Game.Settings.Server.TimestampFormat);
-				writer.WriteLine("[{0}] {1}", timestamp, value);
+				writer.WriteLine("[{0}] {1}", timestamp, item.Text);
 			}
 		}
 
-		public static void Write(string channelName, string format, params object[] args)
+		static IEnumerable<string> FilenamesForChannel(string baseFilename)
 		{
-			var channel = Channel(channelName);
-			if (channel.Writer == null)
-				return;
+			var path = Platform.SupportDir + "Logs";
+			Directory.CreateDirectory(path);
 
-			if (!channel.IsTimestamped)
-				channel.Writer.WriteLine(format, args);
-			else
+			for (var i = 0; ; i++)
+				yield return Path.Combine(path, i > 0 ? "{0}.{1}".F(baseFilename, i) : baseFilename);
+		}
+
+		static ChannelInfo GetChannel(string channelName)
+		{
+			if (!Channels.TryGetValue(channelName, out var info))
+				throw new ArgumentException("Tried logging to non-existent channel " + channelName, nameof(channelName));
+
+			return info;
+		}
+
+		public static void AddChannel(string channelName, string baseFilename, bool isTimestamped = false)
+		{
+			if (Channels.ContainsKey(channelName)) return;
+
+			if (string.IsNullOrEmpty(baseFilename))
 			{
-				var timestamp = DateTime.Now.ToString(Game.Settings.Server.TimestampFormat);
-				channel.Writer.WriteLine("[{0}] {1}", timestamp, format.F(args));
+				Channels.TryAdd(channelName, default);
+				return;
 			}
+
+			foreach (var filename in FilenamesForChannel(baseFilename))
```

Comment:
Style nit: Please add braces (weird to ask this for a change) to avoid the hanging block.